File: problems/tier2-intermediate/015-bookmyshow/solution.js
'use strict';

/* BookMyShow — theater/show/booking + seat locking + contiguous search. */

class ShowOp {
  constructor(kind, s1 = '', s2 = '', s3 = '', s4 = '', i1 = 0, i2 = 0, i3 = 0) {
    this.kind = kind;
    this.s1 = s1;
    this.s2 = s2;
    this.s3 = s3;
    this.s4 = s4;
    this.i1 = i1;
    this.i2 = i2;
    this.i3 = i3;
  }
}

// Seat status constants
const AVAILABLE = 'AVAILABLE';
const LOCKED = 'LOCKED';
const BOOKED = 'BOOKED';

class Seat {
  constructor(row, col) {
    this.row = row;
    this.col = col;
    this.status = AVAILABLE;
    this.lockedBy = '';
    this.lockExpiry = 0;
    this.bookedBy = '';
  }
}

class Show {
  constructor(id, theaterId, movie, time, rows, cols) {
    this.id = id;
    this.theaterId = theaterId;
    this.movie = movie;
    this.time = time;
    this.rows = rows;
    this.cols = cols;
    this.seats = [];
    for (let r = 0; r < rows; r++) {
      const row = [];
      for (let c = 0; c < cols; c++) row.push(new Seat(r, c));
      this.seats.push(row);
    }
  }
}

class Theater {
  constructor(id, name, city) {
    this.id = id;
    this.name = name;
    this.city = city;
  }
}

class BookingSystem {
  constructor() {
    this.theaters = new Map();
    this.shows = new Map();
    this.bookings = new Map();
    this.locks = new Map();
    this.cityMovies = new Map(); // city -> Set of movies
    this.lockCounter = 0;
  }

  _expire_seat(seat, currentTime) {
    if (seat.status === LOCKED && currentTime >= seat.lockExpiry) {
      seat.status = AVAILABLE;
      seat.lockedBy = '';
      seat.lockExpiry = 0;
    }
  }

  addTheater(theaterId, name, city) {
    this.theaters.set(theaterId, new Theater(theaterId, name, city));
  }

  addShow(showId, theaterId, movie, time, rows, cols) {
    this.shows.set(showId, new Show(showId, theaterId, movie, time, rows, cols));
    if (this.theaters.has(theaterId)) {
      const city = this.theaters.get(theaterId).city;
      if (!this.cityMovies.has(city)) this.cityMovies.set(city, new Set());
      this.cityMovies.get(city).add(movie);
    }
  }

  searchMovies(city) {
    if (!this.cityMovies.has(city)) return [];
    // C++ uses std::set which is sorted
    return [...this.cityMovies.get(city)].sort();
  }

  getAvailableSeats(showId, currentTime = 0) {
    if (!this.shows.has(showId)) return [];
    const show = this.shows.get(showId);
    const result = [];
    for (let r = 0; r < show.rows; r++) {
      for (let c = 0; c < show.cols; c++) {
        this._expire_seat(show.seats[r][c], currentTime);
        if (show.seats[r][c].status === AVAILABLE) {
          result.push([r, c]);
        }
      }
    }
    return result;
  }

  bookSeats(bookingId, showId, seatPositions, userId, currentTime = 0) {
    if (!this.shows.has(showId)) return false;
    const show = this.shows.get(showId);
    for (const [r, c] of seatPositions) {
      if (r < 0 || r >= show.rows || c < 0 || c >= show.cols) return false;
      this._expire_seat(show.seats[r][c], currentTime);
      if (show.seats[r][c].status !== AVAILABLE) return false;
    }
    for (const [r, c] of seatPositions) {
      show.seats[r][c].status = BOOKED;
      show.seats[r][c].bookedBy = userId;
    }
    this.bookings.set(bookingId, [bookingId, showId, userId, seatPositions]);
    return true;
  }

  lockSeats(showId, seatPositions, userId, ttlMinutes, currentTime) {
    if (!this.shows.has(showId)) return '';
    const show = this.shows.get(showId);
    for (const [r, c] of seatPositions) {
      if (r < 0 || r >= show.rows || c < 0 || c >= show.cols) return '';
      this._expire_seat(show.seats[r][c], currentTime);
      if (show.seats[r][c].status !== AVAILABLE) return '';
    }
    this.lockCounter += 1;
    const lockId = 'LOCK_' + String(this.lockCounter);
    const expiry = currentTime + ttlMinutes * 60;
    for (const [r, c] of seatPositions) {
      show.seats[r][c].status = LOCKED;
      show.seats[r][c].lockedBy = userId;
      show.seats[r][c].lockExpiry = expiry;
    }
    this.locks.set(lockId, {
      id: lockId,
      showId,
      userId,
      seatPositions,
      expiry,
      confirmed: false,
      released: false,
    });
    return lockId;
  }

  confirmBooking(lockId, currentTime) {
    if (!this.locks.has(lockId)) return false;
    const lock = this.locks.get(lockId);
    if (lock.confirmed || lock.released) return false;
    if (currentTime >= lock.expiry) {
      if (this.shows.has(lock.showId)) {
        const show = this.shows.get(lock.showId);
        for (const [r, c] of lock.seatPositions) {
          this._expire_seat(show.seats[r][c], currentTime);
        }
      }
      lock.released = true;
      return false;
    }
    if (!this.shows.has(lock.showId)) return false;
    const show = this.shows.get(lock.showId);
    for (const [r, c] of lock.seatPositions) {
      show.seats[r][c].status = BOOKED;
      show.seats[r][c].bookedBy = lock.userId;
      show.seats[r][c].lockedBy = '';
      show.seats[r][c].lockExpiry = 0;
    }
    const bookingId = 'BK_' + lockId;
    this.bookings.set(bookingId, [bookingId, lock.showId, lock.userId, lock.seatPositions]);
    lock.confirmed = true;
    return true;
  }

  releaseLock(lockId, currentTime) {
    if (!this.locks.has(lockId)) return false;
    const lock = this.locks.get(lockId);
    if (lock.confirmed || lock.released) return false;
    if (this.shows.has(lock.showId)) {
      const show = this.shows.get(lock.showId);
      for (const [r, c] of lock.seatPositions) {
        show.seats[r][c].status = AVAILABLE;
        show.seats[r][c].lockedBy = '';
        show.seats[r][c].lockExpiry = 0;
      }
    }
    lock.released = true;
    return true;
  }

  findContiguousSeats(showId, n, currentTime = 0) {
    if (!this.shows.has(showId)) return [];
    const show = this.shows.get(showId);
    for (let r = 0; r < show.rows; r++) {
      let count = 0;
      let start = -1;
      for (let c = 0; c < show.cols; c++) {
        this._expire_seat(show.seats[r][c], currentTime);
        if (show.seats[r][c].status === AVAILABLE) {
          if (count === 0) start = c;
          count += 1;
          if (count === n) {
            const res = [];
            for (let i = start; i < start + n; i++) res.push([r, i]);
            return res;
          }
        } else {
          count = 0;
          start = -1;
        }
      }
    }
    return [];
  }
}

function _parse_seats(s) {
  if (!s) return [];
  const out = [];
  for (const token of s.split(';')) {
    if (token.includes(',')) {
      const parts = token.split(',');
      out.push([parseInt(parts[0], 10), parseInt(parts[1], 10)]);
    }
  }
  return out;
}

function show_simulate(ops) {
  const out = [];
  let sys = new BookingSystem();
  let lockSlots = new Array(32).fill('');
  let lastContig = [];
  for (const op of ops) {
    const k = op.kind;
    if (k === 'new') {
      sys = new BookingSystem();
      lockSlots = new Array(32).fill('');
      lastContig = [];
      out.push('ok');
    } else if (k === 'add_theater') {
      sys.addTheater(op.s1, op.s2, op.s3);
      out.push('ok');
    } else if (k === 'add_show') {
      sys.addShow(op.s1, op.s2, op.s3, op.s4, op.i1, op.i2);
      out.push('ok');
    } else if (k === 'movies_count') {
      out.push(String(sys.searchMovies(op.s1).length));
    } else if (k === 'movies_contains') {
      const m = sys.searchMovies(op.s1);
      out.push(m.includes(op.s2) ? 'yes' : 'no');
    } else if (k === 'available_count') {
      out.push(String(sys.getAvailableSeats(op.s1, op.i1).length));
    } else if (k === 'available_has') {
      const v = sys.getAvailableSeats(op.s1, op.i1);
      const found = v.some(p => p[0] === op.i2 && p[1] === op.i3);
      out.push(found ? 'yes' : 'no');
    } else if (k === 'book') {
      const ok = sys.bookSeats(op.s1, op.s2, _parse_seats(op.s3), op.s4, op.i1);
      out.push(ok ? 'ok' : 'fail');
    } else if (k === 'lock') {
      const lid = sys.lockSeats(op.s1, _parse_seats(op.s2), op.s3, op.i1, op.i2);
      if (op.i3 >= 0 && op.i3 < lockSlots.length) lockSlots[op.i3] = lid;
      out.push(lid);
    } else if (k === 'lock_at') {
      out.push(lockSlots[op.i3]);
    } else if (k === 'confirm') {
      const ok = sys.confirmBooking(lockSlots[op.i3], op.i1);
      out.push(ok ? 'ok' : 'fail');
    } else if (k === 'release') {
      const ok = sys.releaseLock(lockSlots[op.i3], op.i1);
      out.push(ok ? 'ok' : 'fail');
    } else if (k === 'release_id') {
      const ok = sys.releaseLock(op.s1, op.i1);
      out.push(ok ? 'ok' : 'fail');
    } else if (k === 'find_contig') {
      lastContig = sys.findContiguousSeats(op.s1, op.i1, op.i2);
      out.push(String(lastContig.length));
    } else if (k === 'contig_at') {
      if (op.i1 < 0 || op.i1 >= lastContig.length) {
        out.push('');
      } else {
        out.push(`${lastContig[op.i1][0]},${lastContig[op.i1][1]}`);
      }
    } else {
      out.push('unknown:' + k);
    }
  }
  return out;
}

module.exports = { ShowOp, show_simulate };

File: problems/tier2-intermediate/015-bookmyshow/solution.py
"""BookMyShow — theater/show/booking + seat locking + contiguous search."""


class ShowOp:
    def __init__(self, kind, s1="", s2="", s3="", s4="", i1=0, i2=0, i3=0):
        self.kind = kind
        self.s1 = s1
        self.s2 = s2
        self.s3 = s3
        self.s4 = s4
        self.i1 = i1
        self.i2 = i2
        self.i3 = i3


# Seat status constants
AVAILABLE = "AVAILABLE"
LOCKED = "LOCKED"
BOOKED = "BOOKED"


class Seat:
    def __init__(self, row, col):
        self.row = row
        self.col = col
        self.status = AVAILABLE
        self.lockedBy = ""
        self.lockExpiry = 0
        self.bookedBy = ""


class Show:
    def __init__(self, id, theaterId, movie, time, rows, cols):
        self.id = id
        self.theaterId = theaterId
        self.movie = movie
        self.time = time
        self.rows = rows
        self.cols = cols
        self.seats = [[Seat(r, c) for c in range(cols)] for r in range(rows)]


class Theater:
    def __init__(self, id, name, city):
        self.id = id
        self.name = name
        self.city = city


class BookingSystem:
    def __init__(self):
        self.theaters = {}
        self.shows = {}
        self.bookings = {}
        self.locks = {}
        self.cityMovies = {}  # city -> set of movies (use sorted set)
        self.lockCounter = 0

    def _expire_seat(self, seat, currentTime):
        if seat.status == LOCKED and currentTime >= seat.lockExpiry:
            seat.status = AVAILABLE
            seat.lockedBy = ""
            seat.lockExpiry = 0

    def addTheater(self, theaterId, name, city):
        self.theaters[theaterId] = Theater(theaterId, name, city)

    def addShow(self, showId, theaterId, movie, time, rows, cols):
        self.shows[showId] = Show(showId, theaterId, movie, time, rows, cols)
        if theaterId in self.theaters:
            city = self.theaters[theaterId].city
            self.cityMovies.setdefault(city, set()).add(movie)

    def searchMovies(self, city):
        if city not in self.cityMovies:
            return []
        # C++ uses std::set which is sorted
        return sorted(self.cityMovies[city])

    def getAvailableSeats(self, showId, currentTime=0):
        if showId not in self.shows:
            return []
        show = self.shows[showId]
        result = []
        for r in range(show.rows):
            for c in range(show.cols):
                self._expire_seat(show.seats[r][c], currentTime)
                if show.seats[r][c].status == AVAILABLE:
                    result.append((r, c))
        return result

    def bookSeats(self, bookingId, showId, seatPositions, userId, currentTime=0):
        if showId not in self.shows:
            return False
        show = self.shows[showId]
        for r, c in seatPositions:
            if r < 0 or r >= show.rows or c < 0 or c >= show.cols:
                return False
            self._expire_seat(show.seats[r][c], currentTime)
            if show.seats[r][c].status != AVAILABLE:
                return False
        for r, c in seatPositions:
            show.seats[r][c].status = BOOKED
            show.seats[r][c].bookedBy = userId
        self.bookings[bookingId] = (bookingId, showId, userId, seatPositions)
        return True

    def lockSeats(self, showId, seatPositions, userId, ttlMinutes, currentTime):
        if showId not in self.shows:
            return ""
        show = self.shows[showId]
        for r, c in seatPositions:
            if r < 0 or r >= show.rows or c < 0 or c >= show.cols:
                return ""
            self._expire_seat(show.seats[r][c], currentTime)
            if show.seats[r][c].status != AVAILABLE:
                return ""
        self.lockCounter += 1
        lockId = "LOCK_" + str(self.lockCounter)
        expiry = currentTime + ttlMinutes * 60
        for r, c in seatPositions:
            show.seats[r][c].status = LOCKED
            show.seats[r][c].lockedBy = userId
            show.seats[r][c].lockExpiry = expiry
        self.locks[lockId] = {
            "id": lockId, "showId": showId, "userId": userId,
            "seatPositions": seatPositions, "expiry": expiry,
            "confirmed": False, "released": False,
        }
        return lockId

    def confirmBooking(self, lockId, currentTime):
        if lockId not in self.locks:
            return False
        lock = self.locks[lockId]
        if lock["confirmed"] or lock["released"]:
            return False
        if currentTime >= lock["expiry"]:
            if lock["showId"] in self.shows:
                show = self.shows[lock["showId"]]
                for r, c in lock["seatPositions"]:
                    self._expire_seat(show.seats[r][c], currentTime)
            lock["released"] = True
            return False
        if lock["showId"] not in self.shows:
            return False
        show = self.shows[lock["showId"]]
        for r, c in lock["seatPositions"]:
            show.seats[r][c].status = BOOKED
            show.seats[r][c].bookedBy = lock["userId"]
            show.seats[r][c].lockedBy = ""
            show.seats[r][c].lockExpiry = 0
        bookingId = "BK_" + lockId
        self.bookings[bookingId] = (bookingId, lock["showId"], lock["userId"], lock["seatPositions"])
        lock["confirmed"] = True
        return True

    def releaseLock(self, lockId, currentTime):
        if lockId not in self.locks:
            return False
        lock = self.locks[lockId]
        if lock["confirmed"] or lock["released"]:
            return False
        if lock["showId"] in self.shows:
            show = self.shows[lock["showId"]]
            for r, c in lock["seatPositions"]:
                show.seats[r][c].status = AVAILABLE
                show.seats[r][c].lockedBy = ""
                show.seats[r][c].lockExpiry = 0
        lock["released"] = True
        return True

    def findContiguousSeats(self, showId, n, currentTime=0):
        if showId not in self.shows:
            return []
        show = self.shows[showId]
        for r in range(show.rows):
            count = 0
            start = -1
            for c in range(show.cols):
                self._expire_seat(show.seats[r][c], currentTime)
                if show.seats[r][c].status == AVAILABLE:
                    if count == 0:
                        start = c
                    count += 1
                    if count == n:
                        return [(r, i) for i in range(start, start + n)]
                else:
                    count = 0
                    start = -1
        return []


def _parse_seats(s):
    if not s:
        return []
    out = []
    for token in s.split(";"):
        if "," in token:
            parts = token.split(",")
            out.append((int(parts[0]), int(parts[1])))
    return out


def show_simulate(ops):
    out = []
    sys = BookingSystem()
    lockSlots = [""] * 32
    last_contig = []
    for op in ops:
        k = op.kind
        if k == "new":
            sys = BookingSystem()
            lockSlots = [""] * 32
            last_contig = []
            out.append("ok")
        elif k == "add_theater":
            sys.addTheater(op.s1, op.s2, op.s3)
            out.append("ok")
        elif k == "add_show":
            sys.addShow(op.s1, op.s2, op.s3, op.s4, op.i1, op.i2)
            out.append("ok")
        elif k == "movies_count":
            out.append(str(len(sys.searchMovies(op.s1))))
        elif k == "movies_contains":
            m = sys.searchMovies(op.s1)
            out.append("yes" if op.s2 in m else "no")
        elif k == "available_count":
            out.append(str(len(sys.getAvailableSeats(op.s1, op.i1))))
        elif k == "available_has":
            v = sys.getAvailableSeats(op.s1, op.i1)
            found = any(p[0] == op.i2 and p[1] == op.i3 for p in v)
            out.append("yes" if found else "no")
        elif k == "book":
            ok = sys.bookSeats(op.s1, op.s2, _parse_seats(op.s3), op.s4, op.i1)
            out.append("ok" if ok else "fail")
        elif k == "lock":
            lid = sys.lockSeats(op.s1, _parse_seats(op.s2), op.s3, op.i1, op.i2)
            if 0 <= op.i3 < len(lockSlots):
                lockSlots[op.i3] = lid
            out.append(lid)
        elif k == "lock_at":
            out.append(lockSlots[op.i3])
        elif k == "confirm":
            ok = sys.confirmBooking(lockSlots[op.i3], op.i1)
            out.append("ok" if ok else "fail")
        elif k == "release":
            ok = sys.releaseLock(lockSlots[op.i3], op.i1)
            out.append("ok" if ok else "fail")
        elif k == "release_id":
            ok = sys.releaseLock(op.s1, op.i1)
            out.append("ok" if ok else "fail")
        elif k == "find_contig":
            last_contig = sys.findContiguousSeats(op.s1, op.i1, op.i2)
            out.append(str(len(last_contig)))
        elif k == "contig_at":
            if op.i1 < 0 or op.i1 >= len(last_contig):
                out.append("")
            else:
                out.append(f"{last_contig[op.i1][0]},{last_contig[op.i1][1]}")
        else:
            out.append("unknown:" + k)
    return out
